feat(worker): Add sourcemap support for worker bundles

KallynGowdy opened the pull request:
<!-- Thank you for contributing! -->

### Description

Improves the [built-in worker plugin](https://github.com/vitejs/vite/blob/d2fc1a1ead1363a15dfcc25fce8587f03289b590/packages/vite/src/node/plugins/worker.ts) to support emitting sourcemaps for generated worker bundles. Closes #5395.

Tests have been added to validate that sourcemaps are generated for inline, non-inline, and shared workers.
<!-- Please insert your description here and provide especially info about the "what" this PR is solving -->

### Additional context

<!-- e.g. is there anything you'd like reviewers to focus on? -->

3 additonal test packages have been added:

-   `sourcemap`
-   `sourcemap-hidden`
-   `sourcemap-inline`

Each of these packages are copied from the `worker` test package and adds a `vite.config.ts` to enable sourcemaps for the respective test case (`true`, `hidden`, and `inline`). The actual test code validates that the `.js.map` files are created and additionally that the source files link to them using a `//# sourceMappingURL=` comment.

One tricky edge case is the scenario where non-inline sourcemaps are being used with inline workers. In this case, the embedded comment has to use the full path (basepath + fileName) instead of simply using the file name. The tests don't address this case, partly because querying the inlined worker code would be pretty messy and it is something where manually running the playground catches the issue pretty easily.

Other details are mentioned via comments in `worker.ts`.

---

### What is the purpose of this pull request? <!-- (put an "X" next to an item) -->

- [ ] Bug fix
- [x] New Feature
- [ ] Documentation update
- [ ] Other

### Before submitting the PR, please make sure you do the following

- [x] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [x] Read the [Pull Request Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md#pull-request-guidelines) and follow the [Commit Convention](https://github.com/vitejs/vite/blob/main/.github/commit-convention.md).
- [x] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`).
- [x] Ideally, include relevant tests that fail without this PR but pass with it.

KallynGowdy committed:
feat(worker): Add sourcemap support for worker bundles

- Supports true, "inline", and "hidden" for config.build.sourcemap options.
- Supports inline and non-inline workers

KallynGowdy committed:
Apply suggestions from code review

Co-authored-by: Shinigami <chrissi92@hotmail.de>

Crisfole commented:
@Shinigami92 It looks like this is working but might be waiting on mac/windows servers to run? And has been for a really long time. What would this take to get moving forward?

Shinigami92 commented:
You need to update the branch with the newest main

KallynGowdy commented:
@Shinigami92 Should I do that?

Shinigami92 commented:
Would be nice :)

Crisfole commented:
@KallynGowdy I just checked: rebase and merge both happen without incident!

KallynGowdy committed:
Merge remote-tracking branch 'origin/main' into feat/worker-sourcemaps

Shinigami92 commented:
uwu :slightly_frowning_face: seems there is an issue with windows?!

Shinigami92 commented:
Mh I will let rerun the tests, could be an issue in CI not your fault :thinking:

KallynGowdy commented:
@Shinigami92 Is there any update on this?

KallynGowdy committed:
Merge remote-tracking branch 'origin/main' into feat/worker-sourcemaps

KallynGowdy committed:
feat(worker): Update recent changes to support sourcemaps

poyoho commented on packages/vite/src/node/plugins/worker.ts:
```diff
@@ -26,16 +32,77 @@ export async function bundleWorkerEntry(
+  }
+
+  return {
+    code: Buffer.from(code)
```
It doesn't seem necessary to use buffer all the time

poyoho replied:
I think it's necessary if it's inline or not 🤔

poyoho commented on packages/vite/src/node/plugins/worker.ts:
```diff
@@ -6,13 +6,19 @@ import type Rollup from 'rollup'
+  sourcemap: Rollup.SourceMap | undefined
+}
+
 export async function bundleWorkerEntry(
```
`bundleWorkerEntry` will always use with `emitSourcemapForWorkerEntry`. In the laster version just diff with params `query`, may be `emitSourcemapForWorkerEntry` can exec in `bundleWorkerEntry`?

poyoho commented:
hi @KallynGowdy. Are you interested in continuing working on it? if not I can fork the PR and let it forward. ❤️

poyoho commented:
> @poyoho Yes, I can update it. Thanks for the offer though!

I think you can try to think about my unit test writing method 😊

KallynGowdy committed:
chore: Merge remote-tracking branch 'vite/main' into feat/worker-sourcemaps

# Conflicts:
#	packages/vite/src/node/plugins/worker.ts
#	packages/vite/src/node/plugins/workerImportMetaUrl.ts

KallynGowdy committed:
feat: Improve bundleWorkerEntry() to emit the sourcemaps on its own

KallynGowdy committed:
refactor: Integrate sourcemap tests into the worker test playground

KallynGowdy committed:
chore: Remove unused test code

KallynGowdy committed:
chore: Remove more unused test code

KallynGowdy committed:
refactor: Simplify integration between bundleWorkerEntry() and emitSourcemapForWorkerEntry()

KallynGowdy committed:
chore: Remove unused interfaces

poyoho reviewed (dismissed):
look good to me. great work here!

KallynGowdy commented:
Looks like another flaky test run...

poyoho commented:
There should be no problem. I often encounter this situation 🙆‍♂️

KallynGowdy committed:
chore: Merge remote-tracking branch 'vite/main' into feat/worker-sourcemaps

# Conflicts:
#	packages/vite/src/node/plugins/worker.ts
#	packages/vite/src/node/plugins/workerImportMetaUrl.ts
